File: 05_Development/CareBridgeMobileApp/lib/integrations/firebaseRealtime/conversation_event_signal.dart
/// Client-side shape of the minimal Firebase RTDB payload (ADR-DCC-004 §1) — exactly the
/// 5 fields the backend ever writes. Signal-only: never trust anything beyond "something
/// changed"; the client always reconciles via REST (`GET /timeline`), never renders this
/// payload directly.
class ConversationEventSignal {
  final String eventId;
  final String eventType;
  final String conversationId;
  final String resourceId;
  final DateTime occurredAt;

  const ConversationEventSignal({
    required this.eventId,
    required this.eventType,
    required this.conversationId,
    required this.resourceId,
    required this.occurredAt,
  });

  factory ConversationEventSignal.fromSnapshotValue(Map<Object?, Object?> raw) {
    return ConversationEventSignal(
      eventId: raw['eventId'] as String,
      eventType: raw['eventType'] as String,
      conversationId: raw['conversationId'] as String,
      resourceId: raw['resourceId'] as String,
      occurredAt: DateTime.fromMillisecondsSinceEpoch(
        (raw['occurredAt'] as num).toInt(),
        isUtc: true,
      ),
    );
  }
}

File: 05_Development/CareBridgeWebApp/src/shared/integrations/firebaseRealtime/conversationEventSignal.ts
/** Client-side shape of the minimal Firestore payload (ADR-DCC-004 §1) — the exact 5 fields
 * the backend ever writes. Signal-only: never trust anything beyond "something changed";
 * the client always reconciles via REST (GET /timeline), never renders this directly. */
export interface ConversationEventSignal {
  eventId: string;
  eventType: string;
  conversationId: string;
  resourceId: string;
  occurredAt: string; // ISO instant, converted from the epoch-millis wire value
}

export function parseConversationEventSignal(raw: unknown): ConversationEventSignal | null {
  if (typeof raw !== 'object' || raw === null) return null;
  const value = raw as Record<string, unknown>;
  if (
    typeof value.eventId !== 'string' ||
    typeof value.eventType !== 'string' ||
    typeof value.conversationId !== 'string' ||
    typeof value.resourceId !== 'string' ||
    typeof value.occurredAt !== 'number'
  ) {
    return null;
  }
  return {
    eventId: value.eventId,
    eventType: value.eventType,
    conversationId: value.conversationId,
    resourceId: value.resourceId,
    occurredAt: new Date(value.occurredAt).toISOString(),
  };
}
